### PR TITLE
Add support for stable diffusion inference on inf2 - `NeuronStableDiffusionPipeline`

### PR DESCRIPTION
**Pull request by JingyaHuang**
- [x] Add `NeuronStableDiffusionPipeline` for SD inference
- [ ] Test of sd pipeline

### TODO
* More tests
* img2img2 and inpaint support

### Snippet

```python
from optimum.neuron import NeuronStableDiffusionPipeline

model_id = "hf-internal-testing/tiny-stable-diffusion-torch"
# model_id = "optimum/tiny-stable-diffusion-neuronx"
input_shapes = {"batch_size": 1, "sequence_length": 18, "height": 64, "width": 64}

stable_diffusion = NeuronStableDiffusionPipeline.from_pretrained(
    model_id, export=True, device_ids=[0, 1], **input_shapes
)
# stable_diffusion = NeuronStableDiffusionPipeline.from_pretrained(model_id, device_ids=[0, 1])
save_directory = "sd_neuron_tiny/"
stable_diffusion.save_pretrained(save_directory)


prompt = "sailing ship in storm by Leonardo da Vinci"
image = stable_diffusion(prompt).images[0]
```

### CODE REVIEW

File: optimum/exporters/neuron/__main__.py [hunk 1]
```diff
@@ -83,10 +92,17 @@ def infer_task(task: str, model_name_or_path: str) -> str:
 
 def normalize_input_shapes(task: str, args: argparse.Namespace) -> Dict[str, int]:
     if task == "stable-diffusion":
-        mandatory_shapes = {
-            name: getattr(args, name, None)
-            for name in getattr(inspect.getfullargspec(build_stable_diffusion_components_mandatory_shapes), "args")
-        }
+        args = vars(args) if isinstance(args, argparse.Namespace) else args
+        mandatory_axes = set(
```
Comment (dacorvo) [inline optimum/exporters/neuron/__main__.py:96]: 🪓 -> axis

File: optimum/neuron/modeling.py [hunk 2]
```diff
@@ -742,7 +742,7 @@ def sample(
             `return_dict_in_generate=True` or a [`~generation.SampleEncoderDecoderOutput`] if
             `model.config.is_encoder_decoder=True`.
 
-        ```"""
+        """
```
Comment (dacorvo) [inline optimum/neuron/modeling.py:745]: Everybody's fixing my mistake ... but I merged the fix I think.

File: optimum/neuron/modeling_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        inputs = (input_ids,)
+        outputs = self.model(*inputs)
+        return outputs
+
+
+class NeuronModelUnet(_NeuronDiffusionModelPart):
+    def __init__(
+        self,
+        model: torch.jit._script.ScriptModule,
+        parent_model: NeuronBaseModel,
```
Comment (dacorvo) [inline optimum/neuron/modeling_diffusion.py:442]: Do you really need the back-reference to the parent_model ? I did not see where it is actually used ...

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+        if prompt is not None and isinstance(prompt, str):
+            batch_size = 1
+        elif prompt is not None and isinstance(prompt, list):
+            batch_size = len(prompt)
+        else:
+            batch_size = prompt_embeds.shape[0]
+
+        if prompt_embeds is None:
+            # textual inversion: procecss multi-vector tokens if necessary
```
Comment (dacorvo) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:56]: process (I assume this comes form the original code).

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                attention_mask = text_inputs.attention_mask.to(device)
+            else:
+                attention_mask = None
+
+            # [Modified] Input and its dtype constraints
+            prompt_embeds = self.text_encoder(input_ids=text_input_ids.to(torch.int32))
+            prompt_embeds = prompt_embeds[0]
+
+        bs_embed, seq_len, _ = prompt_embeds.shape
+        # duplicate text embeddings for each generation per prompt, using mps friendly method
```
Comment (dacorvo) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:91]: mps like in Apple's Metal Performance Shaders ? Maybe we can remove that comment. Unless we want the changes kept to a minimum.

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            elif batch_size != len(negative_prompt):
+                raise ValueError(
+                    f"`negative_prompt`: {negative_prompt} has batch size {len(negative_prompt)}, but `prompt`:"
+                    f" {prompt} has batch size {batch_size}. Please make sure that passed `negative_prompt` matches"
+                    " the batch size of `prompt`."
+                )
+            else:
+                uncond_tokens = negative_prompt
+
+            # textual inversion: procecss multi-vector tokens if necessary
```
Comment (dacorvo) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:116]: process

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+                max_length=max_length,
+                truncation=True,
+                return_tensors="pt",
+            )
+
+            negative_prompt_embeds = self.text_encoder(uncond_input.input_ids.to(torch.int32))
+            negative_prompt_embeds = negative_prompt_embeds[0]
+
+        if do_classifier_free_guidance:
+            # duplicate unconditional embeddings for each generation per prompt, using mps friendly method
```
Comment (dacorvo) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:133]: mps again

File: docs/source/guides/models.mdx [hunk 8]
```diff
@@ -120,4 +120,53 @@ And the next time when you want to run inference, just load your compiled model
 As you see, there is no need to precise shape information and compilation arguments used during the compilation as they are 
 saved in a `config.json` file, and will be restored automatically by `NeuronModelForXXX` class.
 
+## Export and Inference of Generative Models
+
+only add decoder specific items, ex: mandatory shape(bs_size), dependency(transformers-neuronx?)
```
Comment (dacorvo) [inline docs/source/guides/models.mdx:125]: You can just insert a TODO here. I will make the changes later.

**Review by dacorvo — COMMENTED** [review]
Nice ! Only minor comments.

File: optimum/neuron/modeling_diffusion.py (hunk 3 above)
Comment (JingyaHuang) [reply optimum/neuron/modeling_diffusion.py:442]: Oh, it's true because I inherited `NeuronBaseModel` which made it not necessary to have it as an attribute... (I will remove the inheritance and keep the attribute so that I could use some methods of `NeuronBaseModel`  when it makes sense.

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py (hunk 5 above)
Comment (JingyaHuang) [reply optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:91]: Yeah, it comes from the original code in diffusers that I kept them to reduce the potential diff to check... But its true that it's quite irrelevant... Will remove them

File: optimum/exporters/neuron/__main__.py (hunk 1 above)
Comment (dacorvo) [reply optimum/exporters/neuron/__main__.py:96]: Yes, you're right ! Forget about my comment ...

**Review by dacorvo — APPROVED** [review]

File: docs/source/guides/models.mdx [hunk 9]
```diff
@@ -120,4 +120,54 @@ And the next time when you want to run inference, just load your compiled model
...
+
+## Inference of Stable Diffusion Models
+
+Optimum extends 🤗`Diffusers` to support inference on Neuron. To get started, make sure you have installed Diffusers:
+
+```bash
+pip install diffusers >= 0.17.0
+```
+
+You can also accelerate the inference of stable diffusion on neuronx devices (inf / trn1). There are three components which need to be exported to `.neuron` format to boost the performance:
```
Comment (michaelbenayoun) [inline docs/source/guides/models.mdx:135]:
```suggestion
You can also accelerate the inference of stable diffusion on neuronx devices (inf / trn1). There are three components which need to be exported to `.neuron` format to boost the performance:
```
Three or four?

File: docs/source/guides/models.mdx [hunk 10]
```diff
@@ -120,4 +120,54 @@ And the next time when you want to run inference, just load your compiled model
...
+>>> prompt = "a photo of an astronaut riding a horse on mars"
+>>> image = stable_diffusion(prompt).images[0]
+```
+
+<img
+  src="https://raw.githubusercontent.com/huggingface/optimum-neuron/main/docs/assets/guides/models/01-sd-image.png"
+  alt="search ami"
+/>
+
+
```
Comment (michaelbenayoun) [inline docs/source/guides/models.mdx:172]:
```suggestion
### Export with the Optimum CLI
```

Comment (michaelbenayoun) [reply docs/source/guides/models.mdx:172]: Maybe add the equivalent command line to run.

File: optimum/exporters/neuron/__main__.py [hunk 11]
```diff
@@ -83,10 +92,17 @@ def infer_task(task: str, model_name_or_path: str) -> str:
...
+            getattr(inspect.getfullargspec(build_stable_diffusion_components_mandatory_shapes), "args")
+        )
+        # Remove `sequence_length` as diffusers will pad it to the max anyway.
+        mandatory_axes.remove("sequence_length")
+        if not mandatory_axes.issubset(set(args.keys())):
+            raise AttributeError(
+                f"Shape of {mandatory_axes} are mandatory for neuron compilation, while {mandatory_axes.difference(args.keys())} are not given."
+            )
+        mandatory_shapes = {name: args[name] for name in mandatory_axes}
```
Comment (michaelbenayoun) [inline optimum/exporters/neuron/__main__.py:105]: I feel like instead of branching a lot in methods like that, we should have stable diffusion specific functions for that and just call them here.

From my experience with `optimum.exporters.onnx`, adding Stable Diffusion support added a lot of "complexity" in the code because we always need to do something different for these kinds of models. WDTY?

File: optimum/exporters/neuron/model_configs.py [hunk 12]
```diff
@@ -255,35 +258,11 @@ def check_model_inputs_order(
...
-    ):
-        return super().check_model_inputs_order(model=model, dummy_inputs=dummy_inputs, forward_with_tuple=True)
-
-
-@register_in_tasks_manager("unet", *["semantic-segmentation"])
+@register_in_tasks_manager("unet", *["stable-diffusion", "semantic-segmentation"])
 class UNetNeuronConfig(VisionNeuronConfig):
     ATOL_FOR_VALIDATION = 1e-3
+    MANDATORY_AXES = ("batch_size", "sequence_length", "num_channels", "width", "height")
+    model_type = "unet"
```
Comment (michaelbenayoun) [inline optimum/exporters/neuron/model_configs.py:265]: Maybe add this class attribute in `NeuronConfig`.

```
class NeuronConfig:
    model_type: Optional[str] = None
```

File: optimum/exporters/neuron/utils.py [hunk 13]
```diff
@@ -35,42 +67,88 @@
...
 
 
+class DiffusersPretrainedConfig(PretrainedConfig):
+    # override to update `model_type`
+    def to_dict(self):
+        """
+        Serializes this instance to a Python dictionary.
+
+        Returns:
+            :obj:`Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
```
Comment (michaelbenayoun) [inline optimum/exporters/neuron/utils.py:77]:
```suggestion
            :obj:`Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance.
```

File: optimum/exporters/neuron/utils.py [hunk 14]
```diff
@@ -35,42 +67,88 @@
...
+        return output
+
+
+def build_stable_diffusion_components_mandatory_shapes(
+    batch_size: Optional[int] = None,
+    sequence_length: Optional[int] = None,
+    num_channels: Optional[int] = None,
+    height: Optional[int] = None,
+    width: Optional[int] = None,
+    **kwargs,
```
Comment (michaelbenayoun) [inline optimum/exporters/neuron/utils.py:89]: Why?

File: optimum/neuron/modeling_base.py [hunk 15]
```diff
@@ -87,6 +87,13 @@ def __init__(
         self.input_static_shapes = NeuronBaseModel.get_input_static_shapes(self.neuron_config)
         self._attributes_init(model_save_dir, preprocessors, **kwargs)
 
+    @classmethod
+    def _auto_model_to_task(cls, auto_model_class):
+        """
+        Get the task corresponding to a class (for example AutoModelForXXX in transformers).
+        """
+        return TasksManager.infer_task_from_model(auto_model_class)
```
Comment (michaelbenayoun) [inline optimum/neuron/modeling_base.py:95]: Not needed.

File: optimum/neuron/modeling_diffusion.py [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+
+    def _save_pretrained(
+        self,
+        save_directory: Union[str, Path],
+        text_encoder_file_name: str = NEURON_FILE_NAME,
+        unet_file_name: str = NEURON_FILE_NAME,
+        vae_encoder_file_name: str = NEURON_FILE_NAME,
+        vae_decoder_file_name: str = NEURON_FILE_NAME,
+        **kwargs,
```
Comment (michaelbenayoun) [inline optimum/neuron/modeling_diffusion.py:173]: Why?

File: optimum/neuron/modeling_diffusion.py [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+        revision: Optional[str] = None,
+        cache_dir: Optional[str] = None,
+        text_encoder_file_name: Optional[str] = NEURON_FILE_NAME,
+        unet_file_name: Optional[str] = NEURON_FILE_NAME,
+        vae_encoder_file_name: Optional[str] = NEURON_FILE_NAME,
+        vae_decoder_file_name: Optional[str] = NEURON_FILE_NAME,
+        local_files_only: bool = False,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        device_ids: Optional[List[int]] = None,
+        **kwargs,
```
Comment (michaelbenayoun) [inline optimum/neuron/modeling_diffusion.py:230]: Why?

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+class StableDiffusionPipelineMixin(StableDiffusionPipeline):
+    # Adapted from https://github.com/huggingface/diffusers/blob/v0.18.2/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L302
+    def _encode_prompt(
+        self,
+        prompt,
+        device,
+        num_images_per_prompt,
+        do_classifier_free_guidance,
+        negative_prompt=None,
```
Comment (michaelbenayoun) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:38]: Type annotation

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+        prompt_embeds: Optional[torch.FloatTensor] = None,
+        negative_prompt_embeds: Optional[torch.FloatTensor] = None,
+        lora_scale: Optional[float] = None,
+    ):
+        if lora_scale is not None and isinstance(self, LoraLoaderMixin):
+            self._lora_scale = lora_scale
+
+        if prompt is not None and isinstance(prompt, str):
```
Comment (michaelbenayoun) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:46]:
```suggestion
        if isinstance(prompt, str):
```
Is enough here, same comments for similar tests in this function and the next.

File: optimum/neuron/utils/argument_utils.py [hunk 20]
```diff
@@ -172,3 +173,12 @@ def store_compilation_config(
     config_args["output_names"] = output_names
 
     update_func("neuron", config_args)
+
+    if hasattr(config, "_diffusers_version"):
+        import diffusers
+
+        update_func("_diffusers_version", diffusers.__version__)
+
+    model_type = getattr(config, "model_type", None) or model_type
```
Comment (michaelbenayoun) [inline optimum/neuron/utils/argument_utils.py:182]: What if both are `None`?

File: optimum/exporters/neuron/__main__.py (hunk 11 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/__main__.py]: done for this function, maybe we can also separate the main_export of sd, but let's do it when I add other tasks (the moment that the complexity increases so that I can have a better picture).

File: optimum/exporters/neuron/utils.py (hunk 14 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/utils.py:90]: Removed `**kwargs` as it's not necessary

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py (hunk 19 above)
Comment (JingyaHuang) [reply optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:46]: The code comes from the diffusers library that I overridden for neuron, I would prefer to keep it.

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py (hunk 18 above)
Comment (JingyaHuang) [reply optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:38]: hmmm...it turns out that the code overriden also came from diffusers, in this case I would keep it as it is.

**Comment by JingyaHuang** [comment]
Will merge this, and in one or multiple coming PR:
* Tests 
* Add docstring for API doc
* Optimize installation / env setup
* More tasks: img2img / inpaint
...
